libraries/AP_Proximity/AP_Proximity_RangeFinder.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_Proximity_RangeFinder.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <ctype.h>
+#include <stdio.h>
+
+extern const AP_HAL::HAL& hal;
+
+AP_Proximity_RangeFinder::AP_Proximity_RangeFinder(AP_Proximity &_frontend,
+                                   AP_Proximity::Proximity_State &_state) :
+    AP_Proximity_Backend(_frontend, _state)
+{
+}
+
+// update the state of the sensor
+void AP_Proximity_RangeFinder::update(void)
+{
+    // exit immediately if no rangefinder object
+    const RangeFinder *rngfnd = frontend.get_rangefinder();
+    if (rngfnd == nullptr) {
+        set_status(AP_Proximity::Proximity_NoData);
+        return;
+    }
+
+    // look through all rangefinders
+    for (uint8_t i=0; i<rngfnd->num_sensors(); i++) {
+        if (rngfnd->has_data(i)) {
```

Comment:
I think we want to pull in the raw rangefinder distances even if they're too short or too long (i..e  beyond the _MIN_CM or _MAX_CM) but perhaps set the _distance_valid to true only if it's within the acceptable range.  We might also need to handle the situation different depending upon whether the range is too short vs too long.  For example, if it's too short we may still want to move away from it instead of suddenly pretending that there's nothing there.
Anyway, I agree there's an issue here but the solution isn't perfectly clear (to me) at the moment.